pkg/image/controller/controller.go

```diff
@@ -22,78 +24,146 @@ type ImportController struct {
+	imports := make(map[string]api.DockerImageReference)
+	references := util.NewStringSet()
+
+	// read explicitly defined tags
+	for tagName, specTag := range stream.Spec.Tags {
+		if specTag.From != nil {
+			if specTag.From.Kind == "DockerImage" && !specTag.Reference {
+				ref, err := api.ParseDockerImageReference(specTag.From.Name)
+				if err != nil {
+					glog.V(2).Infof("error parsing DokcerImage %s: %v", specTag.From.Name, err)
+					continue
+				}
+				imports[tagName] = ref.DockerClientDefaults()
+			} else {
+				references.Insert(tagName)
+			}
+		}
+	}
+
+	if len(stream.Spec.DockerImageRepository) == 0 {
+		return imports, nil
+	}
+
+	// read tags from default upstream image repository
+	streamRef, err := api.DockerImageReferenceForStream(stream)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := client.Connect(streamRef.Registry, insecure)
+	if err != nil {
+		return nil, err
+	}
+
+	tags, err := conn.ImageTags(streamRef.Namespace, streamRef.Name)
 	switch {
 	case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
-		return c.done(stream, err.Error(), retryCount)
+		return nil, c.done(stream, err.Error(), retryCount)
 	case err != nil:
-		return err
+		return nil, err
 	}
 
-	imageToTag := make(map[string][]string)
 	for tag, image := range tags {
-		if specTag, ok := stream.Spec.Tags[tag]; ok && specTag.From != nil {
-			// spec tag is set to track another tag - do not import
+		if _, ok := imports[tag]; ok {
+			continue
+		}
+		if references.Has(tag) {
 			continue
 		}
 
-		imageToTag[image] = append(imageToTag[image], tag)
-	}
-
-	// no tags to import
-	if len(imageToTag) == 0 {
-		return c.done(stream, "", retryCount)
+		idTagPresent := false
+		for t2, i2 := range tags {
+			if t2 == tag {
+				continue
+			}
+			if i2 == image && t2 == image {
+				idTagPresent = true
+				break
+			}
+		}
+		ref := streamRef
+		if idTagPresent {
+			ref.Tag = image
+		} else {
+			ref.Tag = tag
+		}
+		ref.ID = image
+		imports[tag] = ref
 	}
+	return imports, nil
+}
 
-	for id, tags := range imageToTag {
-		dockerImage, err := conn.ImageByID(ref.Namespace, ref.Name, id)
-		switch {
-		case dockerregistry.IsRepositoryNotFound(err), dockerregistry.IsRegistryNotFound(err):
-			return c.done(stream, err.Error(), retryCount)
-		case dockerregistry.IsImageNotFound(err):
-			continue
-		case err != nil:
-			return err
+// importTags imports tags specified in a map from given ImageStream. Returns an error if one occurs.
+func (c *ImportController) importTags(stream *api.ImageStream, imports map[string]api.DockerImageReference, client dockerregistry.Client, insecure bool) error {
+	retrieved := make(map[string]*dockerregistry.Image)
+	for tag, ref := range imports {
+		var dockerImage *dockerregistry.Image
+		if image, ok := retrieved[ref.ID]; ok {
+			dockerImage = image
+		} else {
+			// TODO insecure applies to the stream's spec.dockerImageRepository, not necessarily to an external one!
```

Comment:
I guess it's not critical. However, users could be confused if they add the insecure annotation to the stream and `--insecure-registry` to `/etc/sysconfig/docker` for the registry/stream in question, but they don't have it for a spec tag ref that pulls from another insecure registry that's not listed in `/etc/sysconfig/docker`. Anyways, all registries should be secured ( :smile: ), so ...